### PR TITLE
`@Flaky`: block flaky tests from CI 

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
We lose a lot of time due to flaky tests, this is also a bad experience for new users

## Approach
Allow us to mark tests as flaky, these are skipped with a message. This may be for a single OS, or for all of them.

❗️Ignoring tests only on CI may be the wrong move, open to discussion here, but wanted the skeleton of the code in

## How Has This Been Tested?
Unit test

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerNoParamTest.kt [hunk 1]
```diff
@@ -32,6 +32,8 @@ import com.ichi2.anki.reviewer.FullScreenMode.Companion.setPreference
 import com.ichi2.anki.reviewer.MappableBinding
 import com.ichi2.libanki.Consts
 import com.ichi2.libanki.DeckId
+import com.ichi2.testutils.Flaky
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/ReviewerNoParamTest.kt:35]: maybe we wanted `Flaky.OS`, was unsure here

File: AnkiDroid/src/test/java/com/ichi2/testutils/IgnoreFlakyTestsInCIRule.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                    else -> throw IllegalStateException("Unknown OS: $osName")
+                }
+            }
+        }
+    }
+}
+
+class IgnoreFlakyTestsTest {
+    @get:Rule
+    val ignoreFlakyTests = IgnoreFlakyTestsInCIRule()
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/testutils/IgnoreFlakyTestsInCIRule.kt:81]: necessary if not an inheritor of `RobolectricTest`

**Review by mikehardy — APPROVED** [review]
I'm okay with all of it, in re: your comments, this is a good change I think.
I hate that reviewer swipe test!

**Comment by oakkitten** [comment]
Could we re-run flaky tests instead?

**Comment by mikehardy** [comment]
I've done that with a retry rule in other repos. End result is that you do execute the test more (local pre-PR plus in CI) but the flakiness is then just ignored and never actually fixed. Not sure it's better in other words. Having the flake exposed locally (vs retry passing) while not blocking CI seems as good a stance as any. I *think* it may be better in fact, and I'm curious to see how it goes with regard to subtle test fixes that otherwise would never happen (retry rule experience) over the next year or so.

Maybe still results in zero subtle test fixes. Maybe results in more.

**Pull request merged**